src/biweight.jl

```diff
@@ -0,0 +1,179 @@
+    return BiweightMidcovariance(c, modify_sample_size)
+end
+
+function covzm(ce::BiweightMidcovariance, x::AbstractVector{<:Real})
+    MADx = median(abs.(x))
+    numerator = zero(eltype(x))
+
+    # If MADx is zero, return zero.
+    MADx == 0 && return zero(one(eltype(x)) / 1)
+
+    denominator = zero(eltype(x))
+    count = 0
+    for xi in x
+        ui² = (xi / (ce.c * MADx))^2
+        ui² >= 1 && continue
+        count += 1
+        numerator += xi^2 * (1 - ui²)^4
+        denominator += (1 - ui²) * (1 - 5 * ui²)
+    end
+    n = ce.modify_sample_size ? count : length(x)
+    return n * numerator / (denominator^2)
+end
+
+function cov(
+    ce::BiweightMidcovariance,
+    x::AbstractVector{<:Real};
+    mean::Union{Nothing,<:Real}=nothing,
+)
+    Mx = isnothing(mean) ? median(x) : mean
+    return covzm(ce, x .- Mx)
+end
+
+function covzm(
+    ce::BiweightMidcovariance, x::AbstractVector{<:Real}, y::AbstractVector{<:Real}
+)
+    MADx = median(abs.(x))
+    MADy = median(abs.(y))
+
+    # Promote types between x & y for numerator
+    numerator = zero(promote_type(eltype(x), eltype(y)))
+
+    # If either of the MADs are zero, return zero.
+    MADx == 0 && return numerator
+    MADy == 0 && return numerator
+
+    denominator_x = zero(eltype(x))
+    denominator_y = zero(eltype(y))
+    count = 0
+    for (xi, yi) in zip(x, y)
+        ui² = (xi / (ce.c * MADx))^2
+        vi² = (yi / (ce.c * MADy))^2
+
+        if ui² < 1
+            denominator_x += (1 - ui²) * (1 - 5 * ui²)
+        end
+
+        if vi² < 1
+            denominator_y += (1 - vi²) * (1 - 5 * vi²)
+        end
+
+        if ui² < 1 && vi² < 1
+            numerator += xi * (1 - ui²)^2 * yi * (1 - vi²)^2
+            count += 1
+        end
+    end
+
+    n = ce.modify_sample_size ? count : length(x)
+    return n * numerator / (denominator_x * denominator_y)
+end
+
+function cov(
+    ce::BiweightMidcovariance, x::AbstractVector{<:Real}, y::AbstractVector{<:Real}
+)
+    return covzm(ce, x .- median(x), y .- median(y))
+end
+
+function cov(
+    ce::BiweightMidcovariance,
+    X::AbstractMatrix{<:Real};
+    mean::Union{Nothing,AbstractVector{<:Real},AbstractMatrix{<:Real}}=nothing,
+    dims::Int=1,
+)
+    dims ∈ (1, 2) || throw(ArgumentError("Argument dims can only be 1 or 2 (given: $dims)"))
+
+    # If the `mean` argument isn't provided, then we centralise with the median.
+    Xc = isnothing(mean) ? X .- median(X; dims=dims) : X .- mean
+
+    # Standardise the orientation of Xc, and ascertain the number of variables.
+    Xc = dims == 2 ? transpose(Xc) : Xc
+    p = size(Xc, 2)
+
+    # Infer the output element type. This follows the approach used in the stdlib, in
+    # Statistics.covzm.
+    T = promote_type(typeof(one(eltype(Xc)) / 1), eltype(Xc))
+
+    # Allocate a concrete output buffer, and populate it by iterating over all pairs.
+    # Since different outliers may be removed
+    result = Matrix{T}(undef, p, p)
```

Comment:
I haven't tried but given you know `T <: Real` here, it's probably better to do `zeros(T, p, p)` rather than go via `undef`

Reply:
What's better about `zeros(T, p, p)`?

Reply:
```
julia> foo() = zeros(Float32, 5_000, 5_000);

julia> bar() = Matrix{Float32}(undef, 5_000, 5_000);

julia> @btime foo(); @btime bar();
  10.683 ms (2 allocations: 95.37 MiB)
  946.000 ns (2 allocations: 95.37 MiB)
```

I learned something today, I somehow thought the former was faster, good to know!